src/Codeception/Lib/InnerBrowser.php

```diff
@@ -1063,11 +1063,19 @@ protected function getFormFor(Crawler $node)
      */
     protected function getFormValuesFor(Form $form)
     {
+        $formNodeCrawler = new Crawler($form->getFormNode());
         $values = [];
         $fields = $form->all();
         foreach ($fields as $field) {
-            if ($field instanceof FileFormField || $field->isDisabled() || !$field->hasValue()) {
-                continue;
+            if ($field instanceof FileFormField || $field->isDisabled()
+                ||
+                !$field->hasValue()
+            ) {
+                // if unchecked a checkboc and if there is hidden input with same name to submit unchecked value
+                $hiddenInput = $formNodeCrawler->filter('input[type=hidden][name="'.$field->getName().'"]');
```

Comment:
Why is this needed? Since we are looping over all form fields I would expect us to find the hidden input eventually, no special action required..
Can you explain?

Reply:
Input type with hidden and with same name as of `coffee` is not selected:

```
<input type="hidden" name="coffee" value="0123">
```

When I iterate over all fields 

```php
$fields = $form->all();
foreach ($fields as $field) {
```

and checkbox is unchecked then I get `NULL` in the value of coffee. Then that null is filtered out by `!$field->hasValue()` in line https://github.com/Codeception/lib-innerbrowser/blob/1.4.1/src/Codeception/Lib/InnerBrowser.php#L1069


So I searched (filtered) explicitly for input type hidden

Reply:
Clear. Is that a bug in the dom crawler then? :wink:

Reply:
Possibly not.


I don't have concrete knowledge regarding your question but there are are 4 actors involved in it

 - HTTP protocol
 - browser implementing that protocol
 - Server software such as Apache, nginx
 - Language like PHP who talks with Apache etc

How above software handle that might differ from software to software

From StackOverflow see answer comment with ~138 upvote - https://stackoverflow.com/a/1992745

So my conclusion is that there is not a bug in the dom crawler

Software tend to grab the last input only. That is why I used `$hiddenInput->last()->attr('value');`

Reply:
I understand your solution, but what I'm unclear on is why it is needed.
The actors you mentioned are not relevant here. We don't use HTTP protocol, server software, a browser or FastCGI.

All we do is parse HTML using Symfony's DOM Crawler. This crawler has a function `->all()` which should return all fields.
If we iterate over these fields we expect to find both the checkbox AND the hidden input.
So if we then go ahead and store these values when the form field is valid we should automatically find the hidden input at some point...

To add to this, the DOM crawler itself has a `getPHPValues()` that already supposedly does the exact same thing. Could you try testing if this function works for our purposes?
That way we remove all manual parsing from this module and instead trust the upstream implementation.

Reply:
I checked `Symfony\Component\DomCrawler\Form::getPHPValues()` . `coffee` is not present there in the returned array. Also I tried `Form::has()` and `Form::get`. In that I get `coffee` but that is of `input type="checkbox"` not of hidden

Reply:
I can confirm that Symfony is struggling with this as well. Currently they can't handle it properly since they store the form field as a map from field name to object... 

Let me think some more on this.